### DISCUSSION TITLE
Vivado Block design challenges

### DISCUSSION THREAD

**Discussion by davidhdeze** [tsfpga/tsfpga#131] [Q&A] [answered]
Hello!

I'm testing TSFPGA flows with the Xilinx Block Designs (BD).

I found some challenges when the BD includes IPs from a user repository (instead of Xilinx default's).

## Build flow

In the build flow I tried a redefinition of the VivadoProject:pre_create() method to include
the user IP repository by means of a `self.tcl_sources.append()` command, analogously as it's done in 
`TsfpgaExampleVivadoProject:pre_create()`

However, this approach lead to a `create_vivado_project.tcl` with tcl calls
into the wrong order:

`module_kcu105_testPrj.py: get_build_projects(self)` piece of code where I create the VivadoProject:
```python

        user_tcl_sources = [
            tcl_dir / "kcu105_pcieDdr_IpExample_bd.tcl", # as created by write_bd_tcl
            tcl_dir / "create_bd_wrapper.tcl"
        ] 
		
        projects.append(
            # VivadoProject(
            MicroblazeBdVivadoProject(
            defined_at=self.path,
            name=self.name,
            modules=modules,
            part=part,
            top="kcu105_pcieDdr_wrapper",
            tcl_sources=user_tcl_sources,
            # .pre_create() user params
            bd_ipUserRepo = Path(tcl_dir / "set_ipRepo.tcl"),
            bd_board = Path(tcl_dir / "set_board.tcl"),
            # .post_build() user parameters of vivado_tsfpga/MicroblazeBdVivadoProject
            elf_src_path = THIS_DIR / "sdk_src",
            elf_app_name = "ipExample_ub_test",
            bd_name = "kcu105_pcieDdr", # TODO: done manually. To be improved
            )
        )
```

`create_vivado_project.tcl` piece of code where .tcls are called

```tcl
# ------------------------------------------------------------------------------
source -notrace {C:/Users/d/anaconda3/Lib/site-packages/tsfpga/vivado/tcl/vivado_default_run.tcl}
source -notrace {C:/Users/d/anaconda3/Lib/site-packages/tsfpga/vivado/tcl/vivado_fast_run.tcl}
source -notrace {C:/Users/d/anaconda3/Lib/site-packages/tsfpga/vivado/tcl/vivado_messages.tcl}
source -notrace {K:/ipExample_axi_wrapper/modules/kcu105_testPrj/tcl/kcu105_pcieDdr_IpExample_bd.tcl}
source -notrace {K:/ipExample_axi_wrapper/modules/kcu105_testPrj/tcl/create_bd_wrapper.tcl}
source -notrace {K:/ipExample_axi_wrapper/modules/kcu105_testPrj/tcl/set_ipRepo.tcl}
source -notrace {K:/ipExample_axi_wrapper/modules/kcu105_testPrj/tcl/set_board.tcl}
```

As you can see, `tcl_sources` are added before the tcl sources included by the pre_create() method so
Vivado executes kcu105_pcieDdr_IpExample_bd.tcl, to create the BD, before setting the IP user repo so it
can't find user IPs.

### Build flow workaround

Not creating a custom pre_create() method but add all tcls as `tcl_sources` in the desired order:

`module_kcu105_testPrj.py: get_build_projects(self)` piece of code where I create the VivadoProject:
```python

        user_tcl_sources = [
            tcl_dir / "set_board.tcl",
            tcl_dir / "set_ipRepo.tcl", # auto created by createIpUserRepo()
            tcl_dir / "kcu105_pcieDdr_IpExample_bd.tcl", # as created by write_bd_tcl
            tcl_dir / "create_bd_wrapper.tcl",
        ]
		
        projects.append(
            # VivadoProject(
            MicroblazeBdVivadoProject(
            defined_at=self.path,
            name=self.name,
            modules=modules,
            part=part,
            top="kcu105_pcieDdr_wrapper",
            tcl_sources=user_tcl_sources,
            # .post_build() user parameters of vivado_tsfpga/MicroblazeBdVivadoProject
            elf_src_path = THIS_DIR / "sdk_src",
            elf_app_name = "ipExample_ub_test",
            bd_name = "kcu105_pcieDdr", # TODO: done manually. To be improved
            )
        ) 
```

`create_vivado_project.tcl` piece of code where .tcls are called

```tcl
# ------------------------------------------------------------------------------
source -notrace {C:/Users/d/anaconda3/Lib/site-packages/tsfpga/vivado/tcl/vivado_default_run.tcl}
source -notrace {C:/Users/d/anaconda3/Lib/site-packages/tsfpga/vivado/tcl/vivado_fast_run.tcl}
source -notrace {C:/Users/d/anaconda3/Lib/site-packages/tsfpga/vivado/tcl/vivado_messages.tcl}
source -notrace {K:/ipExample_axi_wrapper/modules/kcu105_testPrj/tcl/set_board.tcl}
source -notrace {K:/ipExample_axi_wrapper/modules/kcu105_testPrj/tcl/set_ipRepo.tcl}
source -notrace {K:/ipExample_axi_wrapper/modules/kcu105_testPrj/tcl/kcu105_pcieDdr_IpExample_bd.tcl}
source -notrace {K:/ipExample_axi_wrapper/modules/kcu105_testPrj/tcl/create_bd_wrapper.tcl}
```

The workaround is pretty easy and I think VivadoProject class is already enough versatile with 
the tcl_sources parameter.

In case `pre_create()` was the elegant way according to the TSFPA philosophy, VivadoProject.create()
should be changed to execute `.pre_create()` before `self._setup_tcl_sources()`.

## Simulation flow

I updated Vunit package according to the [Vunit issue #962](https://github.com/VUnit/vunit/pull/962)
so the `extract_compile_order.tcl` is able to get BD's sources apart of the IP's.

After that, I've tried to use SimulationProject from `tsfpga.examples.simulation_utils.py`
and, in order hack a bit the system, I've included the BD tcl as is was an IP tcl.

```
modules
├───basic_bd
│   ├───ip_cores
│   │   └───create_bd.tcl
│   └───sim

```

At this point, the simulation flow works properly as long as the BD only has Xilinx IP's,
but it turns more difficult if I want the BD to be create with IPs from a User repository.

The simplest workaround is to include the instructions to add the user ip repository in the `create_bd.tcl`.
This implies to modify by hand the tcl that Vivado generates with `write_bd_tcl` command and to include the 
following lines:

```tcl
# Set User IP repository paths
set repoPath "K:/ipExample_axi_wrapper/modules/kcu105_testPrj/ip_userRepo" 
set obj [get_filesets sources_1]
if { $obj != {} } {
set_property "ip_repo_paths" "[file normalize "$repoPath"]" $obj
# Rebuild user ip_repo's index before adding any source files
update_ip_catalog -rebuild
}
```

That is not a big problem but generating the global path of the repository.
But it turns into a big problem if you don't know the path of IP user repository in forehand,
e.g. if it is generated by another script.

In this case, I have successfully tested another workaround, modifying the Module(BaseModule):get_simulation_files()

```python
class Module(BaseModule):
    def get_simulation_files(
        self,
        include_tests: bool = True,
        files_include: Optional[set[Path]] = None,
        files_avoid: Optional[set[Path]] = None,
        include_vhdl_files: bool = True,
        include_verilog_files: bool = True,
        include_systemverilog_files: bool = True,
        **kwargs: Any,
    ):
        """
        Includes a User IP repository
        """
        
        # Create a TCL for including an IP User Repository in the Vivado prj
        # Put the tcl into the ip_cores folder to 
        ipRepo = createIpUserRepo(
            userRepo = THIS_DIR / "ip_userRepo",
            ipList = [REPO_ROOT, 
                    REPO_ROOT / "lib" / "tpg_axi_wrapper"],
            userIpRepoTcl = THIS_DIR / "ip_cores" / "aaa_set_ipRepo.tcl"
            )

        # call the original get_simulation_files() with the same params
        return super().get_simulation_files(
            include_tests=include_tests,
            files_include=files_include,
            files_avoid=files_avoid,
            include_vhdl_files=include_vhdl_files,
            include_verilog_files=include_verilog_files,
            include_systemverilog_files=include_systemverilog_files,
            **kwargs,
        )
```

The `createIpUserRepo()` creates the IP user repo folder. Additionally, creates the 
"aaa_set_ipRepo.tcl" file inside the "ip_cores" folder so that TSPFA tools interpret it 
as another IP. Note the tricky name of "aaa_*" so it will be alphabetically the first
in the IP files.

Is there a more elegant way to do that in TSPFA?

I know that Xilinx BDs ara a pain and probably should not be supported by TSFPGA.

Thank you very much in advance and congratulations for the great work!

**Comment by LukasVik** [comment, marked as answer]
Hello again @davidhdeze 

# Build flow

I think your workaround is perfect, that's probably how I would do it. 

I would say that using `pre_create` as in `TsfpgaExampleVivadoProject` is best suited for things that should be added to ALL builds, without the user having to specify anything in the `TsfpgaExampleVivadoProject` constructor. If you want to place this functionality in `pre_create` you could do like this:

```python
# Insert first so IP repo paths are set up for all upcoming calls.
self.tcl_sources.insert(0, SOME_DIR / "tcl" / "set_ipRepo.tcl")
```

**Reply by davidhdeze** [reply]
Ok. I understand now the purpose of `pre_create`

**Comment by LukasVik** [comment]
@davidhdeze 

# Simulation flow

A cleaner approach might be to overload the `get_ip_core_files` (https://tsfpga.com/api_reference/tsfpga.html#tsfpga.module.BaseModule.get_ip_core_files) method of the module. 
You could call `createIpUserRepo` there, and add additional IP core files like this:

```python
from tsfpga.ip_core_file import IpCoreFile

....

def get_ip_core_files(
        self,
        **kwargs: Any,
    ) -> list[IpCoreFile]: 
        createIpUserRepo()
        return [IpCoreFile(SOME_DIR / "set_ipRepo.tcl"), IpCoreFile.....] + super().get_ip_core_files(**kwargs)
```

In this case you don't need to do the "aaa" in the file name. 

The call to `super().get_ip_core_files(**kwargs)` will still include everything from the `ip_cores` folder.

**Reply by davidhdeze** [reply]
Great, it works perfectly! Thank you, @LukasVik, for your quick comments and suggestions.

**Comment by LukasVik** [comment]
@davidhdeze Cool, I'm glad it worked out for you!

**Comment by AjSharma-5** [comment]
Hi,

I'm working with a Vivado block design that includes a custom AXI4-Lite peripheral called `muxm_pattern_generator_multi`, connected to the ZYNQ7 Processing System via the M_AXI_GP0 port.

As a beginner in FPGA development, I'm trying to understand how the ZYNQ PS communicates with the IP core. Specifically, I want to clarify:

- How the AXI4-Lite interface allows the PS to send control/data to the custom peripheral.
- What kind of software read/write operations are expected (e.g., using Xil_Out32 in Vitis or SDK).
- How the IP core internally maps those AXI transactions to control logic or registers.
- If there’s an example register map or simple read/write test available.

Any pointers or documentation/examples would be really helpful.

Thanks!

**Reply by LukasVik** [reply]
In general, I don't really have any beginner material for these things. You could start by googling "introduction to axi lite".

In the block design you set up the address space of your M_AXI_GP0 port. A read/write by your embedded software to a physical memory address in that space will map the operation to AXI in your PL. I assume this is what Xil_Out32 does, though I don't know since I stay away from Xilinx code as much as I can. The AXI format of the M_AXI_GP0 port has to be converted to AXI-Lite since that is what most IP (custom or otherwise) use for their register interface.
